src/GHC/Util/Unify.hs

```diff
@@ -0,0 +1,178 @@
+import GHC.Util.HsExpr
+import GHC.Util.Pat
+import GHC.Util.RdrName
+import GHC.Util.View
+
+isUnifyVar :: String -> Bool
+isUnifyVar [x] = x == '?' || isAlpha x
+isUnifyVar [] = False
+isUnifyVar xs = all (== '?') xs
+
+---------------------------------------------------------------------
+-- SUBSTITUTION DATA TYPE
+
+-- A list of substitutions. A key may be duplicated, you need to call
+--  'check' to ensure the substitution is valid.
+newtype Subst' a = Subst' [(String, a)]
+    deriving (Semigroup, Monoid)
+
+-- Unpack the substitution.
+fromSubst' :: Subst' a -> [(String, a)]
+fromSubst' (Subst' xs) = xs
+
+instance Functor Subst' where
+    fmap f (Subst' xs) = Subst' $ map (second f) xs -- Interesting.
+
+instance Outputable a => Show (Subst' a) where
+    show (Subst' xs) = unlines [a ++ " = " ++ unsafePrettyPrint b | (a,b) <- xs]
+
+-- Check the unification is valid and simplify it.
+validSubst' :: (a -> a -> Bool) -> Subst' a -> Maybe (Subst' a)
+validSubst' eq = fmap Subst' . mapM f . groupSort . fromSubst'
+    where f (x, y : ys) | all (eq y) ys = Just (x, y)
+          f _ = Nothing
+
+-- Peform a substition.
+substitute' :: Subst' (LHsExpr GhcPs) -> LHsExpr GhcPs -> LHsExpr GhcPs
+substitute' (Subst' bind) = transformBracketOld' exp . transformBi pat . transformBi typ
+  where
+    exp :: LHsExpr GhcPs -> Maybe (LHsExpr GhcPs)
+    -- Variables.
+    exp (LL _ (HsVar _ x)) = lookup (rdrNameStr' x) bind
+    -- Operator applications.
+    exp (LL loc (OpApp _ lhs (LL _ (HsVar _ x)) rhs))
+      | Just y <- lookup (rdrNameStr' x) bind = Just (cL loc (OpApp noExt lhs y rhs))
+    -- Left sections.
+    exp (LL loc (SectionL _ exp (LL _ (HsVar _ x))))
+      | Just y <- lookup (rdrNameStr' x) bind = Just (cL loc (SectionL noExt exp y))
+    -- Right sections.
+    exp (LL loc (SectionR _ (LL _ (HsVar _ x)) exp))
+      | Just y <- lookup (rdrNameStr' x) bind = Just (cL loc (SectionR noExt y exp))
+    exp _ = Nothing
+
+    pat :: LPat GhcPs -> LPat GhcPs
+    -- Pattern variables.
+    pat (LL _ (VarPat _ x))
+      | Just y@(LL _ HsVar{}) <- lookup (rdrNameStr' x) bind = strToPat' (varToStr' y)
+    pat x = x :: LPat GhcPs
+
+    typ :: LHsType GhcPs -> LHsType GhcPs
+    -- Type variables.
+    typ (LL _ (HsTyVar _ _ x))
+      | Just (LL _ (HsAppType _ _ (HsWC _ y))) <- lookup (rdrNameStr' x) bind = y
+    typ x = x :: LHsType GhcPs
+
+
+---------------------------------------------------------------------
+-- UNIFICATION
+
+type NameMatch' = Located RdrName -> Located RdrName -> Bool
+
+-- | Unification, obeys the property that if @unify a b = s@, then
+-- @substitute s a = b@.
+unify' :: Data a => NameMatch' -> Bool -> a -> a -> Maybe (Subst' (LHsExpr GhcPs))
+unify' nm root x y
+    | Just (x, y) <- cast (x, y) = unifyExp' nm root x y
+    | Just (x, y) <- cast (x, y) = unifyPat' nm x y
+    | Just (x, y) <- cast (x, y) = unifyType' nm x y
+    -- | Just (x :: S) <- cast x = Just mempty -- Confused by this :-/
```

Comment:
@ndmitchell Can you comment on this line please?

Reply:
This is to avoid descending into the source span bit, so the equivalent would be `SrcSpan` in the new world. It ensures that any two src spans match regardless.